packages/snip20/src/query.rs

```diff
@@ -0,0 +1,319 @@
+pub struct Allowance {
+    pub spender: HumanAddr,
+    pub owner: HumanAddr,
+    pub allowance: Uint128,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub expiration: Option<u64>,
+}
+
+/// Balance response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Balance {
+    pub amount: Uint128,
+}
+
+/// Transaction data
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Tx {
+    pub id: u64,
+    pub from: HumanAddr,
+    pub sender: HumanAddr,
+    pub receiver: HumanAddr,
+    pub coins: Coin,
+}
+
+/// TransferHistory response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct TransferHistory {
+    pub txs: Vec<Tx>,
+}
+
+/// Minters response
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Minters {
+    pub minters: Vec<HumanAddr>,
+}
+
+/// SNIP20 queries
+#[derive(Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum QueryMsg<'a> {
+    TokenInfo {},
+    ExchangeRate {},
+    Allowance {
+        owner: &'a HumanAddr,
+        spender: &'a HumanAddr,
+        key: &'a str,
+    },
+    Balance {
+        address: &'a HumanAddr,
+        key: &'a str,
+    },
+    TransferHistory {
+        address: &'a HumanAddr,
+        key: &'a str,
+        page: Option<u32>,
+        page_size: u32,
+    },
+    Minters {},
+}
+
+impl<'a> fmt::Display for QueryMsg<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            QueryMsg::TokenInfo { .. } => write!(f, "TokenInfo"),
+            QueryMsg::ExchangeRate { .. } => write!(f, "ExchangeRate"),
+            QueryMsg::Allowance { .. } => write!(f, "Allowance"),
+            QueryMsg::Balance { .. } => write!(f, "Balance"),
+            QueryMsg::TransferHistory { .. } => write!(f, "TransferHistory"),
+            QueryMsg::Minters { .. } => write!(f, "Minters"),
+        }
+    }
+}
+
+impl<'a> QueryMsg<'a> {
+    /// Returns a StdResult<T>, where T is the "Response" type that wraps the query answer
+    ///
+    /// # Arguments
+    ///
+    /// * `deps` - a reference to the Extern that holds all the external contract dependencies
+    /// * `block_size` - pad the message to blocks of this size
+    /// * `callback_code_hash` - String holding the code hash of the contract being queried
+    /// * `contract_addr` - address of the contract being queried
+    pub fn query<S: Storage, A: Api, Q: Querier, T: DeserializeOwned>(
+        &self,
+        deps: &Extern<S, A, Q>,
+        mut block_size: usize,
+        callback_code_hash: String,
+        contract_addr: HumanAddr,
+    ) -> StdResult<T> {
+        // can not have block size of 0
+        if block_size == 0 {
+            block_size = 1;
+        }
+        let mut msg = to_binary(self)?;
+        space_pad(&mut msg.0, block_size);
+        deps.querier
+            .query(&QueryRequest::Wasm(WasmQuery::Smart {
```

Comment:
I think that instead of requiring the entire `Extern` object, you can require just the `Q` (`: Querier`) object inside. It would simplify the function declaration, and users will be able to see what parts of the system it uses, since they'll have to pass in `&deps.querier` rather than the entire `&deps`. It's a little bit more verbose but i thinks it's better

Reply:
similarly in `handle.rs`